lib/scholar/linear/ridge.ex

```diff
@@ -0,0 +1,324 @@
+
+    {kernel, y} =
+      if opts[:sample_weights_flag] do
+        y = y * Nx.new_axis(sample_weights, 1)
+        {kernel * Nx.outer(sample_weights, sample_weights), y}
+      end
+
+    if one_alpha do
+      kernel = kernel + Nx.eye(num_samples) * alpha[[0]]
+      dual_coeff = Nx.LinAlg.solve(kernel, y)
+
+      if opts[:sample_weights_flag],
+        do: dual_coeff * Nx.new_axis(sample_weights, 1),
+        else: dual_coeff
+    else
+      y = Nx.transpose(y)
+
+      y = Nx.new_axis(y, -1)
+
+      kernel = Nx.new_axis(kernel, 0) |> Nx.broadcast({num_targets, num_samples, num_samples})
+
+      reg =
+        Nx.new_axis(Nx.eye(num_samples), 0)
+        |> Nx.broadcast({num_targets, num_samples, num_samples})
+
+      reg = reg * Nx.reshape(alpha, {:auto, 1, 1})
+      kernel = kernel + reg
+      dual_coeff = Nx.LinAlg.solve(kernel, y) |> Nx.squeeze(axes: [-1])
+
+      dual_coeff =
+        if opts[:sample_weights_flag],
+          do: dual_coeff * Nx.new_axis(sample_weights, 0),
+          else: dual_coeff
+
+      Nx.transpose(dual_coeff)
+    end
+  end
+
+  defnp solve_cholesky(x, y, alpha) do
+    num_features = Nx.axis_size(x, 1)
+    num_targets = Nx.axis_size(y, 1)
+
+    kernel = Nx.dot(x, [0], x, [0])
+
+    xy = Nx.dot(x, [0], y, [0])
+
+    one_alpha = Nx.all(alpha[[0]] == alpha)
+
+    {kernel, xy}
+
+    if one_alpha do
+      kernel = kernel + Nx.eye(num_features) * alpha[[0]]
+      Nx.LinAlg.solve(kernel, xy) |> Nx.transpose()
+    else
+      target = Nx.transpose(xy)
+
+      target = Nx.new_axis(target, -1)
+
+      kernel = Nx.new_axis(kernel, 0) |> Nx.broadcast({num_targets, num_features, num_features})
+
+      reg =
+        Nx.new_axis(Nx.eye(num_features), 0)
+        |> Nx.broadcast({num_targets, num_features, num_features})
+
+      reg = reg * Nx.reshape(alpha, {:auto, 1, 1})
+      kernel = kernel + reg
+      Nx.LinAlg.solve(kernel, target) |> Nx.squeeze(axes: [-1])
+    end
+  end
+
+  defn solve_svd(x, y, alpha) do
+    {u, s, vt} = Nx.LinAlg.svd(x)
+    min_shape = Kernel.min(Nx.axis_size(u, -1), Nx.axis_size(vt, -1))
+    u = Nx.slice_along_axis(u, 0, min_shape, axis: -1)
+    vt = Nx.slice_along_axis(vt, 0, min_shape, axis: -2)
+    s_size = Nx.size(s)
+    alpha_size = Nx.size(alpha)
+    idx = (s > 1.0e-15) |> Nx.new_axis(1) |> Nx.broadcast({s_size, alpha_size})
+    s = Nx.new_axis(s, 1) |> Nx.broadcast({s_size, alpha_size})
+    alpha = Nx.new_axis(alpha, 0) |> Nx.broadcast({s_size, alpha_size})
+    d = Nx.broadcast(0.0, {s_size, alpha_size})
+    d = Nx.select(idx, s / (s ** 2 + alpha), d)
+    uty = Nx.dot(u, [0], y, [0])
+    d_uty = d * uty
+    Nx.dot(vt, [0], d_uty, [0]) |> Nx.transpose()
+  end
+
+  defnp set_intercept(coeff, x_offset, y_offset, fit_intercept?) do
+    if fit_intercept? do
+      y_offset - Nx.dot(coeff, x_offset)
+    else
+      0.0
+    end
+  end
+
+  defn preprocess_data(x, y, sample_weights, opts) do
+    if opts[:sample_weights_flag],
+      do:
```

Comment:
We can add support for full_matrices: false in SVD :)

The code is basically already there, we just need to avoid one or two Nx.concatenate calls!

Reply:
Absolutely, I thought about this PR.